### PR TITLE
Add support to JsonSerializationWriter.WriteNonParsableObjectValue for anonymous objects.

### LINKED ISSUE #306: JsonSerializationWriter.WriteNonParsableObjectValue does not support anonymous objects
I was passing values into the AdditionalData property on the type generated in my Kiota client and I noticed that anonymous object values were being passed as empty objects.

```
{ "myAdditionalDataProperty": { } }
```

From debugging I can see that here https://github.com/microsoft/kiota-dotnet/blob/42ea74edf45d473462dd050661b1972ce14c15e1/src/serialization/json/JsonSerializationWriter.cs#L447 T comes through as "Object" so get properties returns no results. 

No error is thrown, the properties just never show up. It's possible this affects other types but I only tested with anonymous objects.

### PR DESCRIPTION
**Pull request by greg-ww**
Fixes #306.

### CODE REVIEW

File: src/serialization/json/JsonSerializationWriter.cs
Comment (baywet) [inline src/serialization/json/JsonSerializationWriter.cs:475]: why this change?

File: src/serialization/json/JsonSerializationWriter.cs
Comment (baywet) [inline src/serialization/json/JsonSerializationWriter.cs:386]:
```suggestion
                writer.WriteStartObject();
                foreach(DictionaryEntry entry in values)
                {
                    if (entry.Key is not string keyStr)
                        throw new InvalidOperationException($"error serialization additional data value with key {key}, unsupported type {type}");
                    WriteAnyValue(keyStr, entry.Value);
                }
```

While functional, advanced type reflection has a high runtime cost and is going to trip AOT/trimming

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for the contributions!
A couple of remarks, also please sign the CLA (see the bot comment)

**Comment by greg-ww** [comment]
@microsoft-github-policy-service agree company="Civitai"

File: src/serialization/json/JsonSerializationWriter.cs
Comment (greg-ww) [reply src/serialization/json/JsonSerializationWriter.cs:475]: I believe this is covered in the issue #306, but typeof(T) comes back as object and therefore no properties are ever returned by GetProperties(). For a clear demonstration of the issue you can revert this change and run the WriteSampleObjectValue test I modified.

File: src/serialization/json/JsonSerializationWriter.cs
Comment (greg-ww) [reply src/serialization/json/JsonSerializationWriter.cs:386]: Happy to iterate on this.
Is there something specific you would rather I avoid?

Comment (baywet) [reply src/serialization/json/JsonSerializationWriter.cs:386]: yes, we should avoid using reflection as much as possible here since it throws AOT/trimming off and is costly at runtime.
I believe the suggestion I've made is functionally equivalent to the initial proposal here.

Comment (greg-ww) [reply src/serialization/json/JsonSerializationWriter.cs:386]: Ah sorry I missed the suggestion, yeah that works for me.

Comment (greg-ww) [reply src/serialization/json/JsonSerializationWriter.cs:386]: We still need to use reflection for the error message but this is inline with the WriteAnyValue so I assume that's fine.

Comment (greg-ww) [reply src/serialization/json/JsonSerializationWriter.cs:386]: I will look at adding a test for the exception too.

File: src/serialization/json/JsonSerializationWriter.cs
Comment (baywet) [reply src/serialization/json/JsonSerializationWriter.cs:474]: Thanks for the additional information. I'd like @andrueastman to provide his input here. I think this should be fine because of the annotation, but would like a second opinion here.

Comment (andrueastman) [reply src/serialization/json/JsonSerializationWriter.cs:474]: Yeah, this change is okay with me really.
The codebase was always using `value.GetType()` but the use of `typeof(T)` snuck in recently in this PR https://github.com/microsoft/kiota-dotnet/pull/291/files when we were trying to iron out AOT/trimming warnings with the test project. So, this is probably looking to fix a regression that we didn't catch.

**Review by andrueastman — COMMENTED** [review]
Thanks for the contribution @greg-ww 

Any chance you can do the following as well?

- Run `dotnet format` to fix linting errors
- Bump the patch version at https://github.com/microsoft/kiota-dotnet/blob/1eca71b6e4c86f8da5db1c420d1a3ba2cd71ab06/Directory.Build.props#L4
- Add a changelog entry at https://github.com/microsoft/kiota-dotnet/blob/main/CHANGELOG.md

**Comment by greg-ww** [comment]
Will try to implement this feedback today.

File: src/serialization/json/JsonSerializationWriter.cs
Comment (greg-ww) [reply src/serialization/json/JsonSerializationWriter.cs:369]: Yeah fair enough, will do.

**Comment by greg-ww** [comment]
@andrueastman I made the requested modifications, let me know if I missed anything.

**Review by andrueastman — APPROVED** [review]
Looks good to me @greg-ww 

Just need a final review from @baywet before we merge.

**Review by baywet — APPROVED** [review]

**Pull request merged**
